File: leetecode/RemoveNthNode.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        ls = []
        current = head
        while current:
            ls.append(current)
            current = current.next
        if((n+1)>len(ls)):
            if(-n+1>=0):
                return None
            return ls[-n+1]
        if(-n+1>=0):
            ls[-2].next = None
            return head
        ls[-n-1].next = ls[-n].next
        return head

File: leetecode/RemoveNthNode.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }

    ListNode(int x, ListNode next) {
        val = x;
        this.next = next;
    }
}

public class RemoveNthNode {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode nhead = new ListNode(0);
        nhead.next = head;
        ListNode first = nhead;
        ListNode second = head;
        while(n>0)
        {
            second = second.next;
            n--;
        }
        while (second!=null) {
            second = second.next;
            first = first.next;
        }
        first.next = first.next.next;
        return nhead.next;
    }   
}
